src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,272 @@
+    );
+
+    return {
+      events: stream.events.slice(0, maxEventIndex(expectedStreamVersion)),
+      currentStreamVersion: stream.events.length,
+      streamExists: true,
+    };
+  }
+
+  async aggregateStream<State, EventType extends Event>(
+    streamName: StreamName,
+    options: AggregateStreamOptions<State, EventType, number>,
+  ): Promise<AggregateStreamResult<State, number>> {
+    const stream = await this.readStream<EventType>(streamName, options?.read);
+    const state = stream.events.reduce(options.evolve, options.initialState());
+    return {
+      state,
+      currentStreamVersion: stream.currentStreamVersion,
+      streamExists: stream.streamExists,
+    };
+  }
+
+  async appendToStream<EventType extends Event>(
+    streamName: StreamName,
+    events: EventType[],
+    options?: AppendToStreamOptions<number> & {
+      /**
+       * These will be ran after a the events have been successfully appended to
+       * the stream. `appendToStream` will return after every projection is completed.
+       */
+      projections?: Array<Projection<EventType>>;
+    },
+  ): Promise<AppendToStreamResult<number>> {
+    let stream = await this.collection.findOne({
+      streamName: { $eq: streamName },
+    });
+    let currentStreamPosition = stream?.events.length ?? 0;
+    let createdNewStream = false;
+
+    if (!stream) {
+      const result = await this.collection.insertOne({
+        streamName,
+        events: [],
+        createdAt: new Date(),
+        updatedAt: new Date(),
+      });
+      stream = await this.collection.findOne({
+        _id: result.insertedId,
+      });
+      createdNewStream = true;
+    }
+
+    const eventCreateInputs: ReadEvent[] = [];
+    for (const event of events) {
+      currentStreamPosition++;
+      eventCreateInputs.push({
+        type: event.type,
+        data: event.data,
+        metadata: {
+          now: new Date(),
+          eventId: uuid(),
+          streamName,
+          streamPosition: BigInt(currentStreamPosition),
+          ...(event.metadata ?? {}),
+        },
+      });
+    }
+
+    // TODO: better error here, should rarely happen if ever
+    // if another error was not thrown before this
+    if (!stream) throw new Error('Failed to create stream');
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      options?.expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    // @ts-expect-error The actual `EventType` is different across each stream document,
+    // but the collection was instantiated as being `EventStream<Event>`. Unlike `findOne`,
+    // `findOneAndUpdate` does not allow a generic to override what the return type is.
+    const updatedStream: WithId<EventStream<EventType>> | null =
+      await this.collection.findOneAndUpdate(
+        {
+          streamName: { $eq: streamName },
+          events: { $size: stream.events.length },
+          updatedAt: new Date(),
```

Comment:
**CONCERN:** Why do we have this date condition here?

Reply:
I think that I'd then just update it and keep in metadata.

Reply:
Oop, actually, this was a typo.

I _meant_ to make the `updatedAt` as part of the update, not as part of the condition. My bad.

Reply:
I can keep these timestamps if we you think they'd be useful but really this info can be found from the first/last events too so probably just redundant.

Reply:
That's why I was surprised, and why we need automated test 😅👌

Reply:
> I can keep these timestamps if we you think they'd be useful but really this info can be found from the first/last events too so probably just redundant.

I'm fine with having that also on the stream level. It's not taking much additional data, and we can use Mongo `project` not to read it, but just update it.